### PR TITLE
agent: add test coverage for functions find_process and online_resources

### LINKED ISSUE #4085: agent: Add tests for find_process function
Add tests for the `find_process` function in `agent/src/sandbox.rs`. Relates to #293

### LINKED ISSUE #4136: agent: Add tests for online_resources function
Add tests for the `online_resources` function in `agent/src/sandbox.rs`. Relates to #293

### PR DESCRIPTION
**Pull request by Shensd**
Add test coverage for the functions `find_process` and `online_resources` in `src/sandbox.rs`.

This pull request supersedes #4084 

Fixes #4085
Fixes #4136

Signed-off-by: Jack Hance <jack.hance@ndsu.edu>

### CODE REVIEW

**Comment by liubin** [comment]
Hi @Shensd You need to run cargo fmt and push again.

**Comment by Shensd** [comment]
I applied the format changes for both tests to the second commit, I figured this would be the best way to go about this as the second commit already depends on the first. If there is a better way for me to have gone about this let me know and I will make changes accordingly.

**Comment by liubin** [comment]
> I applied the format changes for both tests to the second commit, I figured this would be the best way to go about this as the second commit already depends on the first. If there is a better way for me to have gone about this let me know and I will make changes accordingly.

Normally separated commits should be compiler-ok, I'd recommending fix the first commit and add the test case in the second commit, or squash two commits into one. Either of them, you need to do a `git rebase`.

File: src/agent/src/sandbox.rs [hunk 1]
```diff
@@ -851,4 +852,241 @@ mod tests {
...
     }
+
+    #[tokio::test]
+    async fn test_find_process() {
+        let logger = slog::Logger::root(slog::Discard, o!());
+        let mut s = Sandbox::new(&logger).unwrap();
+        let (mut linux_container, _root) = create_linuxcontainer();
+
+        // add process with pid of 99
+        let test_pid = 99;
```
Comment (jodh-intel) [inline src/agent/src/sandbox.rs:863]: `libc::pid_t` is actually unsigned, so assuming black-box testing it might be good to test some boundary values too such as `std::i32::MIN`, `-1`, `0`, `1`, `std::i32::MAX` in case there are any conversions happening in `find_process()` or the functions it calls.

File: src/agent/src/sandbox.rs [hunk 2]
```diff
@@ -851,4 +852,241 @@ mod tests {
...
+            number_autogen_directories: u32,
+
+            extra_directories: &'a [TestDirectory<'a>],
+            pattern: String,
+            to_enable: i32,
+
+            result: i32,
+        }
+
+        impl Default for TestData<'_> {
```
Comment (jodh-intel) [inline src/agent/src/sandbox.rs:930]: If you change the `struct`s to be `#[derive(Debug, Default)]`, you can delete the `impl Default` block.

File: src/agent/src/sandbox.rs [hunk 3]
```diff
@@ -851,4 +852,241 @@ mod tests {
...
+        #[derive(Debug)]
+        struct TestData<'a> {
+            directory_autogen_name: String,
+            number_autogen_directories: u32,
+
+            extra_directories: &'a [TestDirectory<'a>],
+            pattern: String,
+            to_enable: i32,
+
+            result: i32,
```
Comment (jodh-intel) [inline src/agent/src/sandbox.rs:927]: Could you change this to `Result<i32>` since that would:

- match exactly what `online_resources()` returns
- make it clear reading through `TestData` which tests are expected to pass/fail since you can specify a passing value as `Ok(0)`, `Ok(99)`, _etc_ and failing values as `Err(anyhow!("something bad happened"))`.

**Review by jodh-intel — COMMENTED** [review]
Thanks @Shensd. This is quite a tricky one so great job on getting this working. Just a few comments...

**Review by fidencio — COMMENTED** [review]
@Shensd,

Please, see the comments in the diff.

**Comment by Shensd** [comment]
I added _most_ of the requested changes, the only one I wasn't able to fully remove was the `impl Default` for TestData in `test_online_resources`, as `#[derive(Default)]` wasn't playing nice with `Result<i32>`. 

Otherwise I have rebased and squashed the commits, I threw `cargo fmt --test` and ` cargo clippy --all-targets --all-features --release -- -D warnings` against it and it all looked good, so hopefully I covered everything.

File: src/agent/src/sandbox.rs [hunk 4]
```diff
@@ -847,4 +848,259 @@ mod tests {
...
+
+            let mut test_process = Process::new(
+                &logger,
+                &oci::Process::default(),
+                "this_is_a_test_process",
+                true,
+                1,
+            )
+            .unwrap();
+            // processes interally only have pids when manually set
```
Comment (fidencio) [inline src/agent/src/sandbox.rs:870]: typo: interally -> internally

File: src/agent/src/sandbox.rs [hunk 5]
```diff
@@ -847,4 +848,259 @@ mod tests {
...
+                    pattern: Default::default(),
+                    to_enable: Default::default(),
+                    result: Ok(Default::default()),
+                }
+            }
+        }
+
+        let tests = &[
+            // 4 well formed directories, request enabled 4,
+            // correct result 4 enabled, should pass
```
Comment (fidencio) [inline src/agent/src/sandbox.rs:946]: I really like how you describe what's expected to happen when adding the test data, kudos! :-)

**Review by fidencio — APPROVED** [review]
lgtm, thanks @Shensd!

**Comment by fidencio** [comment]
@jodh-intel, has @Shensd addressed all your comments?

**Review by jodh-intel — APPROVED** [review]
Thanks @Shensd!

lgtm
